test/token/ERC20WithPermit.test.js

```diff
@@ -0,0 +1,1102 @@
+                    )
+                ).to.be.revertedWith("Transfer amount exceeds allowance")
+              })
+            })
+
+            context("when the token owner does not have enough balance", () => {
+              const amount = initialSupply
+
+              beforeEach(async () => {
+                await token
+                  .connect(initialHolder)
+                  .transfer(anotherAccount.address, 1)
+              })
+
+              it("should revert", async () => {
+                await expect(
+                  token
+                    .connect(anotherAccount)
+                    .transferFrom(
+                      initialHolder.address,
+                      recipient.address,
+                      amount
+                    )
+                ).to.be.revertedWith("Transfer amount exceeds allowance")
+              })
+            })
+
+            context("when the token owner is the zero address", () => {
+              const allowance = initialSupply
+
+              it("should revert", async () => {
+                await expect(
+                  token
+                    .connect(anotherAccount)
+                    .transferFrom(ZERO_ADDRESS, recipient.address, allowance)
+                ).to.be.revertedWith("Transfer amount exceeds allowance")
+              })
+            })
+          }
+        )
+      })
+
+      context("when the recipient is the zero address", () => {
+        const allowance = initialSupply
+
+        beforeEach(async () => {
+          await token
+            .connect(initialHolder)
+            .approve(anotherAccount.address, allowance)
+        })
+
+        it("should revert", async () => {
+          await expect(
+            token
+              .connect(anotherAccount)
+              .transferFrom(initialHolder.address, ZERO_ADDRESS, allowance)
+          ).to.be.revertedWith("Transfer to the zero address")
+        })
+      })
+
+      context("when given the maximum allowance", () => {
+        const allowance = MAX_UINT256
+
+        beforeEach(async () => {
+          await token
+            .connect(initialHolder)
+            .approve(anotherAccount.address, allowance)
+        })
+
+        it("should not reduce approved amount", async () => {
+          expect(
+            await token.allowance(initialHolder.address, anotherAccount.address)
+          ).to.equal(allowance)
+
+          await token
+            .connect(anotherAccount)
+            .transferFrom(initialHolder.address, recipient.address, to1e18(100))
+
+          expect(
+            await token.allowance(initialHolder.address, anotherAccount.address)
+          ).to.equal(allowance)
+        })
+      })
+    })
+  })
+
+  describe("approve", () => {
+    context("when the spender is not the zero address", () => {
+      context("when the sender has enough balance", () => {
+        const allowance = initialSupply
+
+        it("should emit an approval event", async () => {
```

Comment:
I know it's validated in `"should approve the requested amount and replaces the previous one"` in another test but can we add `allowance` map check as well under `"when the sender has enough balance"`?